### DISCUSSION TITLE
Interest on onnx2torch?

### DISCUSSION THREAD

**Discussion by GeorgeS2019** [xamarin/TorchSharp#1111] [Ideas]
Just wonder if there is interest on **onnx2torchsharp**
https://github.com/ENOT-AutoDL/onnx2torch

**Comment by GeorgeS2019** [comment]
Potentially relevant projects

- https://github.com/MyCaffe/OnnxControl
- https://github.com/nietras/OnnxSharp
- https://github.com/onnxplorer/ONNXplorer/tree/master/Assets/Scripts/Onnx
- https://github.com/fel88/Dendrite

**Comment by NiklasGustafsson** [comment]
Does this tool generate source code, or in-memory constructs? If someone wanted to do a TorchSharp version, I would suggest spitting out source code rather than using something like Reflection.Emit()

**Reply by GeorgeS2019** [reply]
@NiklasGustafsson 
I think the community is  NOW more inspired compared to previous years. 
This onnx2torchsharp will bring more existing onnx-generated-by-pytorch projects to torchsharp. 

Hopefully, one of the community members is inspired to try.

**Comment by GeorgeS2019** [comment]
Follow the discussion [here](https://github.com/fel88/Dendrite/issues/11#issuecomment-1762849941)

**Comment by GeorgeS2019** [comment]
@NiklasGustafsson 

It is now possible to do training using c# Onnx API.

It would mean it is possible to download exported Onnx from Pytorch from e.g. HuggingFace to continue training in c#

However, there are still many challenges ahead that are potentially commonly shared between TorchSharp and Onnx c# API trainings

One of these is how to have a DOTNET Foundation approach to address [ContribOperators](https://github.com/microsoft/onnxruntime/blob/main/docs/ContribOperators.md).

As this issue will have impact on the future of **ML.NET**, **Torchsharp** and **OnnxRuntime** training and inference using c#

**Comment by GeorgeS2019** [comment]
[Reference](https://github.com/LaurentMazare/tch-rs/issues/695#issuecomment-1529008255)
> **Exporting models to onnx is supported in libtorch** so this would probably be easy to add but I don't think there is a good way to do the opposite: running/importing onnx models from the c++ side, see this https://github.com/pytorch/pytorch/issues/21683. 

> Exporting models to onnx is supported in libtorch so this would probably be easy to add but I don't think there is a good way to do the opposite: running/importing onnx models from the c++ sideI don't think there is a current plan to do this. If the goal is just to retrieve the weight and not the architecture, I would suggest writing some one off python script using the onnx python package and writing some safetensors file. If the goal is to generate a full jit/torch-script model, there seems to be various python libraries doing this but I'm not sure how good they actually are.

> models with torchscript are slower than the onnx version, so I would like to be able to run inference models with the onnx format. But I hear you, thank you.
